### ISSUE TITLE
35/WAKU2-NOISE 37/WAKU2-NOISE-SESSIONS 43/WAKU2-NOISE-PAIRING

### ISSUE THREAD

**Issue by fryorcraken** [waku-org/js-waku#950]
This is a **bug report/feature or change request/support request**

## Problem

There are no documented or proposed way to create a secure session between two nodes.

## Proposed Solutions

Implement https://github.com/vacp2p/rfc/pull/531

## Acceptance Criteria

- [ ] https://github.com/vacp2p/rfc/pull/531 is implemented
- [ ] #681 functionality is available
- [ ] This is used in an Eth-PM-like example.

## Notes

This needs support for https://rfc.vac.dev/spec/35/ and https://rfc.vac.dev/spec/37/
Implementation of these two specs should be planned and track in this issue.

**Comment by fryorcraken** [comment]
@s1fr0 if we want to do an example with nwaku, to create a session channel between js-waku and nwaku, how would it look like? I guess we would need to build a small binary that enables 1:1 chats?

**Comment by fryorcraken** [comment]
The way I'd see it would be:

Browser<>Browser:

1. `https://examples.waku.org/pairing-js` shows a QR code. data is URL A `https://examples.waku.org/pairing-js?<necessary crypto information>`
2. User opens `https://examples.waku.org/pairing-js` on desktop browser
3. User scans QR code from mobile open URL A
4. js-waku in user's mobile browser starts handshake session with user's desktop browser
5. handshake done, user can exchange encrypted chat messages between both devices

Browser<>nwaku

1. `https://examples.waku.org/pairing-js` shows a `<necessary crypto information>` in some easy copy-pastable format (single line string)
2. User opens `https://examples.waku.org/pairing-js` on desktop browser, copies `<necessary crypto information>`
3. User opens `chat2` (or other binary) and pass the `<necessary crypto information>`
4. `chat2` does device pairing with user's desktop browser
8. handshake done, user can exchange encrypted chat messages between both devices

So yes, `chat2` could work but I MO it should be a `private message` only mode.

**Comment by richard-ramos** [comment]
Code for browser <> browser pairing is available at: https://github.com/waku-org/js-noise

The following actions are still pending:
1. Instead of connecting to a hardcoded node, to just connect to the prod fleet instead
2. Add minimal chat example like the one on js-rln.
3. Move the example to js-waku-examples
4. We are currently exporting everything in index.js. It probably makes sense to create separate exports.
5. Add docs to functions

**Comment by richard-ramos** [comment]
In this PR https://github.com/waku-org/js-noise/pull/11 the following items were implemented:
1.  Example connects to the test fleet
2. Add minimal chat example like the one on js-rln.

I'll follow your suggestion of having everything imported from `index.js`, so IMO the only things pending are adding documentation to the functions/readme, and moving the example to js-waku-examples.  (Of course, PRs to improve code org / style / logic are always welcome :) )

**Comment by richard-ramos** [comment]
I'm currently attempting to implement a js-waku <> go-waku example as I have created a pairing object with similar functionality to js-noise in go-waku

**Comment by fryorcraken** [comment]
Todo:
- [ ] Finish the go-waku example as the js-waku is now published at https://examples.waku.org/

The waku socket PoC can be tacked in the existing issue.

**Comment by fryorcraken** [comment]
@weboko can we track the improvement of the existing noise API to replace the `sender` and `responder` API with more generic waku interfaces `ISender` and `IReceiver` implemented by light push, relay and filter.

**Comment by weboko** [comment]
> @weboko can we track the improvement of the existing noise API to replace the `sender` and `responder` API with more generic waku interfaces `ISender` and `IReceiver` implemented by light push, relay and filter.

I made this issue to track it https://github.com/waku-org/js-noise/issues/14

**Comment by fryorcraken** [comment]
@richard-ramos what is the state of the go-waku example?

**Comment by richard-ramos** [comment]
Go-waku's example is finished and pointing to https://examples.waku.org/

**Comment by fryorcraken** [comment]
> Go-waku's example is complete and pointing to [examples.waku.org](https://examples.waku.org/)

Can't compile unfortunately (with v0.5.0):

```
▶ make
go build -o build/noise .
# noise
./chat2.pb.go:26:17: undefined: proto.ProtoPackageIsVersion4
./main.go:39:28: cannot use context.Background() (value of type context.Context) as type node.WakuNodeOption in argument to node.New
./main.go:49:12: not enough arguments in call to wakuNode.Start
    have ()
    want (context.Context)
make: *** [Makefile:4: build] Error 2
```

Next steps:
- [ ] Fix compilation issue in go-waku
- [ ] Add noise binary as part of go-waku release (@richard-ramos, wdyt?)
- [ ] Add text in https://examples.waku.org/noise-js to explain how to download/compile/use the go-waku example

**Comment by fryorcraken** [comment]
Also:

- [ ] Need to update noise example to use latest js-noise as encoding URL as changed.

**Comment by weboko** [comment]
I think now those are done: 
- [x] Fix compilation issue in go-waku
- [x] Need to update noise example to use latest js-noise as encoding URL as changed.
- [ ] Add text in https://examples.waku.org/noise-js to explain how to download/compile/use the go-waku example - https://github.com/waku-org/js-waku-examples/pull/224

Is there any necessity for this one, @fryorcraken ? 
- [ ] Add noise binary as part of go-waku release

**Comment by fryorcraken** [comment]
> Is there any necessity for this one, @fryorcraken ?
> Add noise binary as part of go-waku release

Yes so that one can just download the bin to play with device pairing.

@richard-ramos can you please handle that?

**Comment by weboko** [comment]
Ice boxed till js-noise is back in progress.
